runtime/coverage.go

```diff
@@ -539,3 +552,25 @@ func (r *CoverageReport) MarshalLCOV() ([]byte, error) {
 
 	return buf.Bytes(), nil
 }
+
+// Given a common.Location, returns its mapped source, if any.
+// Defaults to the location's ID().
+func (r *CoverageReport) locationSource(location common.Location) string {
```

Comment:
Maybe always use the ID as the "location source", instead of stripping parts of it (i.e. the kind), which could lead to clashes (e.g. `import Crypto` vs `import "Crypto"`)

Reply:
As far as I understand, we cannot use the `ID`, because the `flow.json` config file contains only contract names, e.g.:
```json
{
  "contracts": {
    "ApprovalVoting": {
      "source": "contracts/ApprovalVoting.cdc",
      "aliases": {
        "testing": "0000000000000007"
      }
    }
  }
}
```
So, if we are dealing with address, string or identifier location, the `location.ID()` would return:
- `A.0000000000000006.ApprovalVoting`
- `S.ApprovalVoting`
- `I.ApprovalVoting`

but the `locationMappings` field, only contains the `"ApprovalVoting"` key. That's why to strip parts of it, and get only the contract name from a location.

Reply:
I have added a test case for identifier locations, which wasn't handled previously: https://github.com/onflow/cadence/commit/4871d50d108d715a582746143ff09d72e78c525b

Reply:
I guess a different way to phrase the same question is, if the flow.json is something like below:
```json
{
  "contracts": {
    "ApprovalVoting": {
      "source": "contracts/ApprovalVoting.cdc",
      "aliases": {
        "testing": "0000000000000007"
      }
    }
  }
}
```

then should all of the following import formats work (are they all refer to the same `ApprovalVoting` in flow.json), with respect to the coverage ?
- `import ApprovalVoting`
- `import "ApprovalVoting"`
- `import ApprovalVoting from 0000000000000007`

Reply:
Yes, they work indeed, and I have added test cases for all three location types:
- [IdentifierLocation](https://github.com/onflow/cadence/pull/2859/files#diff-e3df797dc0e65b26af99a2311291d53b5202848e6ac9b83cdfab3009d0723c3cR707-R716) - `import ApprovalVoting`
- [StringLocation](https://github.com/onflow/cadence/pull/2859/files#diff-e3df797dc0e65b26af99a2311291d53b5202848e6ac9b83cdfab3009d0723c3cR679-R688) - `import "ApprovalVoting"`
- [AddressLocation](https://github.com/onflow/cadence/pull/2859/files#diff-e3df797dc0e65b26af99a2311291d53b5202848e6ac9b83cdfab3009d0723c3cR641-R660) - `import ApprovalVoting from 0000000000000007`

Let me know if I understood the question correctly :sweat_smile:

Reply:
hmm, not sure if `import ApprovalVoting` should actually work (does importing like that actually work in the emulator? i.e: just simple importing, without code coverage or anything). Identifier imports are for built-in contracts. e.g: `import Crypto` or `import Test`.

Reply:
If the emulator already supports identifier imports to be resolved from `flow.json` (like above), then this looks good to me. But might be good to double-check.

Reply:
You are indeed right :+1: 
The emulator does not support importing custom identifier locations. But identifier imports for built-in contracts, do produce code coverage (https://github.com/onflow/cadence-tools/blob/master/test/test_runner.go#L454-L456), that's why these locations are excluded in the `cadence-tools/test` repository.
I can remove the case for handling identifier locations, but it doesn't really have any effect at this point.

Reply:
ah ok, makes sense 👍 
Thanks for the clarification